Initial side navigation implementation for Developer Journey

nbhuiyan opened the pull request:
This is the initial set of changes to have a basic
side navigation feature for devjourney. It does everything
that clicking on the table of content items do. The side
navigation bar remains fixed while scrolling. Also,
the top navigation bar has also been set to remain fixed,
making navigation between pages easier.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

I have been unable to publish it so far due to github pages build error. However, it does work locally. @arianneb Please see if you face the same problem while publishing through gh pages.

Here is a screenshot of how it looks like:

<img width="719" alt="screen shot 2018-04-23 at 6 41 44 pm" src="https://user-images.githubusercontent.com/23691212/39156876-226ef2f2-4726-11e8-98a9-0911fdb7d3e8.png">

nbhuiyan committed:
Initial side navigation implementation for Developer Journey

This is the initial set of changes to have a basic
side navigation feature for devjourney. It does everything
that clicking on the table of content items do. The side
navigation bar remains fixed while scrolling. Also,
the top navigation bar has also been set to remain fixed,
making navigation between pages easier.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan commented:
@rwy0717 @arianneb I have been able to figure out why my website was not publishing. It was a very stupid reason as I am new to gh pages. Here is the website: https://nbhuiyan.github.io/b9/

rwy7 commented:
Very nice! I'm happy accepting this PR as-is, but here are some thoughts on future patches:
* The sidebar should collapse somehow if the user is on their phone. 
* The sidebar should be generated from either the document itself, or if not possible, a yaml data file.
* The sidebar should use scrollspy to follow the user through headings.
* I'm not 100% on the top bar sticking to the top of the screen, especially on the phone. It's very fat. Maybe we can slim it down, or display it only on an upwards scroll?

rwy7 commented:
Actually, can you get rid of the old TOC in this PR?

arianneb commented:
A few comments:
- when you click on an item, the top nav bar hides the heading, so it's not immediately apparent that you've gone to the right place. can we fix this?
- can we make it a little wider?
- as Robert suggested, please delete the original TOC
- the side bar isn't auto generated from the headings? @nbhuiyan 

Otherwise, looks super good!

nbhuiyan committed:
Get rid of Table of Content in Dev Journey

With an initial side navigation bar to jump across sections,
we can now get rid of the TOC for the Dev Journey.

Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

nbhuiyan commented:
Thanks for the feedback. I have removed the TOC at the top. You may view the current state of changes here: https://nbhuiyan.github.io/b9/BuildARuntime.html

While this is just an initial set of changes to have a working side navigation bar, I had a lot in mind that I wanted to implement, but didn't because all of this is new to me. I thought that having a basic sidebar to work on and improve in an iterative manner would be better than trying to do all of it at once. 

To address @rwy0717 's comments:

> The sidebar should collapse somehow if the user is on their phone.

I agree. It could be a major inconvenience to have it in the way on smaller screens.
> The sidebar should be generated from either the document itself, or if not possible, a yaml data file.

Yes, I agree.
> The sidebar should use scrollspy to follow the user through headings.

That is going to be a really neat feature!
> I'm not 100% on the top bar sticking to the top of the screen, especially on the phone. It's very fat. Maybe we can slim it down, or display it only on an upwards scroll?

I did that since I was having issues with the sidebar not filling the entire height when scrolling past the top. Slimming it down would be good, but I am not sure how much work it is going to be to make it appear when scrolling up.

To address @arianneb 's comments:

> when you click on an item, the top nav bar hides the heading, so it's not immediately apparent that you've gone to the right place. can we fix this?

Yes, I noticed the issue and was trying to figure out how to fix it using page anchors. I left it unfixed for this PR since we first needed to settle on whether we wanted to have the top bar fixed or not.
> can we make it a little wider?

I kept it narrow to make it easier to deal with on mobile displays. We can make it wider when we make it collapsible.

> as Robert suggested, please delete the original TOC

Done

> the side bar isn't auto generated from the headings?

Not yet.

I will open an issue to track these.
